### DISCUSSION TITLE
Profiler extensions?

### DISCUSSION THREAD

**Discussion by Trass3r** [microsoft/vscode-discussions#1525] [Extension Development QnA] [answered]
Hi I've been wondering to what extent performance profiling extension are supported.
I saw there's support for JS: https://code.visualstudio.com/docs/nodejs/profiling#_using-the-record-button
But I guess that doesn't extend easily to other languages?

And for showing expensive lines of code in the editor maybe the test coverage API could be shoehorned into that: https://code.visualstudio.com/api/extension-guides/testing#test-coverage
Though it's focused on unit tests and not designed in a generic way.
Also found this: https://github.com/microsoft/vscode/issues/129198

Is there anything else I've missed so far?

**Comment by isidorn** [comment]
I think you could do a profiler extensions for any extension, and contribute custom commands that trigger your experience.
Have you tried this? Have you hit any issues?

**Reply by robconery** [reply]
Hi @Trass3r did this help your issue?

**Reply by Trass3r** [reply]
Not really, but I guess the answer is you have to implement everything yourself? Like displaying sample hit counts in the editor etc.

**Reply by robconery** [reply]
@isidorn is this the case?

**Reply by isidorn** [reply, marked as answer]
That is my understanding yes. Your extension has to implement everything. Test coverage is usually built upon an extension that is contributing testing. And profilling on top of a debugger extension.

**Reply by Trass3r** [reply]
Do you have some hints at relevant APIs or reference code?
Like for displaying a number on each line, or coloring lines etc
